data/transactions/logic/assembler.go

```diff
@@ -0,0 +1,1112 @@
+func (ops *OpStream) Bytes() (program []byte, err error) {
+	var scratch [binary.MaxVarintLen64]byte
+	prebytes := bytes.Buffer{}
+	// TODO: configurable what version to compile for in case we're near a version boundary?
+	version := ops.Version
+	if version == 0 {
+		//version = config.Consensus[protocol.ConsensusCurrentVersion].LogicSigVersion
+		version = AssemblerDefaultVersion
+	}
+	vlen := binary.PutUvarint(scratch[:], version)
+	prebytes.Write(scratch[:vlen])
+	if len(ops.intc) > 0 {
+		prebytes.WriteByte(0x20) // intcblock
+		vlen := binary.PutUvarint(scratch[:], uint64(len(ops.intc)))
+		prebytes.Write(scratch[:vlen])
+		for _, iv := range ops.intc {
+			vlen = binary.PutUvarint(scratch[:], iv)
+			prebytes.Write(scratch[:vlen])
+		}
+	}
+	if len(ops.bytec) > 0 {
+		prebytes.WriteByte(0x26) // bytecblock
+		vlen := binary.PutUvarint(scratch[:], uint64(len(ops.bytec)))
+		prebytes.Write(scratch[:vlen])
+		for _, bv := range ops.bytec {
+			vlen = binary.PutUvarint(scratch[:], uint64(len(bv)))
+			prebytes.Write(scratch[:vlen])
+			prebytes.Write(bv)
+		}
+	}
+	if prebytes.Len() == 0 {
+		program = ops.Out.Bytes()
+		return
+	}
+	pbl := prebytes.Len()
+	outl := ops.Out.Len()
+	out := make([]byte, pbl+outl)
+	pl, err := prebytes.Read(out)
+	if pl != pbl || err != nil {
+		err = fmt.Errorf("wat: %d prebytes, %d to buffer? err=%s", pbl, pl, err)
+		return
+	}
+	ol, err := ops.Out.Read(out[pl:])
+	if ol != outl || err != nil {
+		err = fmt.Errorf("%d program bytes but %d to buffer. err=%s", outl, ol, err)
+		return
+	}
+	program = out
+	return
+}
+
+// AssembleString takes an entire program in a string and assembles it to bytecode
+func AssembleString(text string) ([]byte, error) {
+	sr := strings.NewReader(text)
+	ops := OpStream{}
+	err := ops.Assemble(sr)
+	if err != nil {
+		return nil, err
+	}
+	return ops.Bytes()
+}
+
+type disassembleState struct {
+	program       []byte
+	pc            int
+	out           io.Writer
+	labelCount    int
+	pendingLabels map[int]string
+
+	nextpc int
+	err    error
+}
+
+func (dis *disassembleState) putLabel(label string, target int) {
+	if dis.pendingLabels == nil {
+		dis.pendingLabels = make(map[int]string)
+	}
+	dis.pendingLabels[target] = label
+}
+
+type disassembleFunc func(dis *disassembleState)
+
+type disassembler struct {
+	name    string
+	handler disassembleFunc
+}
+
+var disassemblers = []disassembler{
+	{"intcblock", disIntcblock},
+	{"intc", disIntc},
+	{"bytecblock", disBytecblock},
+	{"bytec", disBytec},
+	{"arg", disArg},
+	{"txn", disTxn},
+	{"gtxn", disGtxn},
+	{"global", disGlobal},
+	{"bnz", disBnz},
+	{"load", disLoad},
+	{"store", disStore},
+}
+
+var disByName map[string]disassembler
+
+func init() {
+	disByName = make(map[string]disassembler)
+	for _, dis := range disassemblers {
+		disByName[dis.name] = dis
+	}
+}
+
+var errTooManyIntc = errors.New("intcblock with too many items")
+
+func parseIntcblock(program []byte, pc int) (intc []uint64, nextpc int, err error) {
```

Comment:
As discussed offline, I think it would make sense to move these functions into another file. It feels weird when we fix a panic in `Check` and the fix ends up in `assembler.go`